File: backend/app/app/models/umn_class.py
import datetime
import re
import uuid
from typing import Any, Dict, List

DAYS_OF_WEEK = ["MONDAY", "TUESDAY", "WEDNESDAY", "THURSDAY", "FRIDAY"]
WEEKDAY_DICT = {
    "MONDAY": 0,
    "TUESDAY": 1,
    "WEDNESDAY": 2,
    "THURSDAY": 3,
    "FRIDAY": 4,
    "SATURDAY": 5,
    "SUNDAY": 6,
}


class UMNClass:
    """
    A class to hold the data for each umn class. Can be serialized to a gcal event
    dictionary.
    """

    def __init__(
        self,
        name: str = "",
        dept: str = "",
        course_num: str = "",
        section: str = "",
        location: str = "",
        start_date_str: str = "",
        end_date_str: str = "",
        start_time_str: str = "",
        end_time_str: str = "",
        days_of_week: List[str] = [],
    ) -> None:
        self.name = name
        self.dept = dept
        self.course_num = course_num
        self.section = section
        self.location = location

        now = datetime.datetime.now()
        self.start_time_str = start_time_str
        self.start_time = (
            datetime.datetime.strptime(start_time_str, "%H:%M:%S").time()
            if start_time_str
            else now.time()
        )
        self.end_time_str = end_time_str
        self.end_time = (
            datetime.datetime.strptime(end_time_str, "%H:%M:%S").time()
            if end_time_str
            else now.time()
        )
        self.start_date_str = start_date_str
        self.start_date = (
            datetime.datetime.strptime(start_date_str, "%Y-%m-%d").date()
            if start_date_str
            else now.date()
        )
        self.end_date_str = end_date_str
        self.end_date = (
            datetime.datetime.strptime(end_date_str, "%Y-%m-%d").date()
            if end_date_str
            else now.date()
        )

        self.days_of_week = days_of_week.copy()

    def __str__(self) -> str:
        """
        Create a string representation of the class
        """
        time_start = self.start_time.strftime("%I:%M %p")
        time_end = self.end_time.strftime("%I:%M %p")
        return (
            f"{self.name}\n"
            + f"{self.dept} {self.course_num} {self.section}\n"
            + f"{self.location}\n"
            + f"{time_start} - {time_end}\n"
            + f"{self.start_date} - {self.end_date}\n"
            + f'Days: {", ".join(self.days_of_week)}\n'
        )

    def __repr__(self) -> str:
        """
        Override the unique representation for the class
        """
        return f"{{{self.dept} {self.course_num} {self.section}}}"

    def __eq__(self, other: object) -> bool:
        """
        Custom equality override for set comparators
        """
        if isinstance(other, UMNClass):
            return (
                self.dept == other.dept
                and self.course_num == other.course_num
                and self.section == other.section
            )
        return False

    def __hash__(self) -> int:
        """
        Custom hash method for membership comparison for sets
        """
        return hash(self.dept + self.course_num + self.section)

    def set_times(self) -> None:
        """
        Convert all the string times into datetime.time objects
        """
        am_or_pm = self.end_time_str[-2:]

        end_datetime = datetime.datetime.strptime(self.end_time_str, "%I:%M%p")
        start_datetime = datetime.datetime.strptime(
            self.start_time_str + am_or_pm, "%I:%M%p"
        )

        # Special case when class start time is am and end time is pm
        if am_or_pm == "PM" and start_datetime > end_datetime:
            start_datetime -= datetime.timedelta(hours=12)

        self.start_time = start_datetime.time()
        self.end_time = end_datetime.time()

        assert self.start_time < self.end_time

    def set_class_dates(
        self, start_date: datetime.date, end_date: datetime.date
    ) -> None:
        """
        Assigns the dates on the class. End date will not be adjusted since google
        calendar handles the offset issues.
        """
        assert isinstance(start_date, datetime.date)
        assert isinstance(end_date, datetime.date)

        start_weekday = start_date.weekday()
        first_weekday = WEEKDAY_DICT[self.days_of_week[0]]
        # See if any of the other days of the week are on the first week
        for day in self.days_of_week:
            if WEEKDAY_DICT[day] >= start_weekday:
                first_weekday = WEEKDAY_DICT[day]
                break
        if first_weekday >= start_weekday:
            offset = first_weekday - start_weekday
        else:
            offset = first_weekday - start_weekday + 7

        self.start_date = start_date + datetime.timedelta(days=offset)
        # Set every class on this end date because gcal will handle the offset
        self.end_date = end_date

    def serialize(self) -> Dict[str, str]:
        """
        Turns the class into a serializable dictionary for json conversion. Generates a
        unique id based on the department, course number, and section
        """
        return {
            "id": f"{self.dept}{self.course_num}{self.section}",
            "name": self.name,
            "dept": self.dept,
            "course_num": self.course_num,
            "section": self.section,
            "location": self.location,
            "start_time": self.start_time.isoformat(),
            "end_time": self.end_time.isoformat(),
            "start_date": self.start_date.isoformat(),
            "end_date": self.end_date.isoformat(),
            "days_of_week": ", ".join(self.days_of_week),
        }

    def to_event_dict(self) -> Dict[str, Any]:
        """
        Turns the class into a recurring gcal event dictionary as specified by
        the google calendar api. This is a json format conforming to the
        icalendar specification.
        """
        start = datetime.datetime.combine(self.start_date, self.start_time)
        end = datetime.datetime.combine(self.start_date, self.end_time)

        # Setup recurrence
        r_days = ",".join([day[:2] for day in self.days_of_week])
        r_end = datetime.datetime.combine(
            self.end_date + datetime.timedelta(days=1), self.end_time
        ).isoformat()
        r_end = re.compile(r"[-:]").sub("", r_end) + "Z"
        recurrence = f"RRULE:FREQ=WEEKLY;BYDAY={r_days};UNTIL={r_end}"

        # Make the event dictionary
        event = {
            "summary": self.name,
            "location": self.location,
            "description": f"{self.dept} {self.course_num} ({self.section})",
            "start": {"dateTime": start.isoformat(), "timeZone": "America/Chicago"},
            "end": {"dateTime": end.isoformat(), "timeZone": "America/Chicago"},
            "recurrence": [recurrence],
        }
        return event

    def to_ics_event_string(self) -> str:
        """
        Converts the class into a string in the icalendar format to be written to an
        ics file:

        BEGIN:VEVENT
        DTSTAMP: *now*
        DTSTART;TZID=America/Chicago:*start time*
        DTEND;TZID=America/Chicago:*end time*
        RRULE:*recurrence rule*
        UID:*uuid*
        DESCRIPTION:*description*
        LOCATION:*location*
        SEQUENCE:0
        STATUS:CONFIRMED
        SUMMARY:*summary*
        TRANSP:OPAQUE
        END:VEVENT
        """
        data = self.to_event_dict()
        start = (
            data["start"]["dateTime"].replace("-", "").replace(":", "").replace(" ", "")
        )
        end = data["end"]["dateTime"].replace("-", "").replace(":", "").replace(" ", "")
        return (
            "BEGIN:VEVENT\n"
            + f'DTSTAMP:{datetime.datetime.now().strftime("%Y%m%dT%H%M%S")}Z\n'
            + f'DTSTART;TZID={data["start"]["timeZone"]}:{start}\n'
            + f'DTEND;TZID={data["end"]["timeZone"]}:{end}\n'
            + f'RRULE:{data["recurrence"][0][6:]}\n'
            + f"UID:{uuid.uuid4()}\n"
            + f'DESCRIPTION:{data["description"]}\n'
            + f'LOCATION:{data["location"]}\n'
            + "SEQUENCE:0\n"
            + "STATUS:CONFIRMED\n"
            + f'SUMMARY:{data["summary"]}\n'
            + "TRANSP:OPAQUE\n"
            + "END:VEVENT\n"
        )

File: extension/src/UMNClass.ts
import { addDays, format, getDay, parse } from 'date-fns';
import CalendarEvent from './CalendarEvent';

const getHours = (timeStr: string) => {
  const match = timeStr.match(/(\d{1,2}):\d{2}/g);
  if (!match) throw new Error('Could not parse time for class');
  else return parseInt(match[0]);
};

const weekdayStringToNumber = (weekdayStr: string) => {
  return ['SUNDAY', 'MONDAY', 'TUESDAY', 'WEDNESDAY', 'THURSDAY', 'FRIDAY', 'SATURDAY'].indexOf(
    weekdayStr
  );
};

export class UMNClass {
  public name: string;
  public dept: string;
  public courseNum: string;
  public section: string;
  public location: string;
  public startDate: Date;
  public endDate: Date;
  public startTimeStr: string;
  public endTimeStr: string;
  public daysOfWeek: Array<string> = [];

  constructor(
    name = '',
    dept = '',
    courseNum = '',
    section = '',
    location = '',
    startDateStr = '',
    endDateStr = '',
    startTimeStr = '',
    endTimeStr = '',
    daysOfWeek: Array<string> = []
  ) {
    this.name = name;
    this.dept = dept;
    this.courseNum = courseNum;
    this.section = section;
    this.location = location;

    this.startDate = parse(startDateStr, 'yyyy-MM-dd', new Date());
    this.endDate = parse(endDateStr, 'yyyy-MM-dd', new Date());
    this.startTimeStr = startTimeStr;
    this.endTimeStr = endTimeStr;
    this.daysOfWeek = [...daysOfWeek];
  }

  /**
   * Returns a human-readable string for debugging purposes.
   *
   * @returns human-readable string for debugging purposes.
   */
  public toString(): string {
    return `UMNClass(${this.name}, ${this.dept} ${this.courseNum} ${this.section}, ${
      this.location
    }, ${this.startTimeStr} to ${this.endTimeStr}, ${this.daysOfWeek.join(',')})`;
  }

  public toEvent(): CalendarEvent {
    const start = parse(
      `${format(this.startDate, 'yyyy-MM-dd')} ${this.startTimeStr}`,
      'yyyy-MM-dd h:mmaa',
      this.startDate
    );
    const end = parse(
      `${format(this.endDate, 'yyyy-MM-dd')} ${this.endTimeStr}`,
      'yyyy-MM-dd h:mmaa',
      this.endDate
    );

    // Setup recurrence
    const recurrenceDays = this.daysOfWeek.map((d) => d.slice(0, 2)).join(',');
    const recurrenceEnd = addDays(end, 1).toISOString().replace(/[:-]/g, '');
    const recurrence = `RRULE:FREQ=WEEKLY;BYDAY=${recurrenceDays};UNTIL=${recurrenceEnd}`;

    // Make the recurrence object
    return {
      summary: this.name,
      location: this.location,
      description: `${this.dept} ${this.courseNum} ${this.section}`,
      start: { dateTime: start.toISOString(), timeZone: 'America/Chicago' },
      end: { dateTime: end.toISOString(), timeZone: 'America/Chicago' },
      recurrence: [recurrence],
    };
  }

  /**
   * Returns a unique hashable code for the class.
   *
   * @returns hash code for the class.
   */
  public hashCode(): string {
    return `${this.name} | ${this.dept}${this.courseNum}${this.section} | ${this.location} | ${this.startTimeStr} to ${this.endTimeStr}`;
  }

  /**
   * Checks for equality with another UMNClass
   *
   * @param other Other UMNClass to compare against
   * @returns true if the classes are the same
   */
  public equals(other?: UMNClass): boolean {
    if (!other) return false;
    return this.toString() === other.toString();
  }

  /**
   * Sets the true course dates based on the input term start and end dates and the
   * days of the week the class is on.
   *
   * @param termStartDate
   * @param termEndDate
   */
  public setTrueCourseDates(termStartDate: Date, termEndDate: Date): void {
    // Get the day of week, 0 representing Sunday to 6 representing Saturday
    const startWeekdayOfTerm = getDay(termStartDate);
    let firstWeekdayOfCourse = weekdayStringToNumber(this.daysOfWeek[0]);

    for (const day of this.daysOfWeek) {
      const weekdayNumber = weekdayStringToNumber(day);
      // See if any of the other days of the week are on the first week
      if (weekdayNumber >= startWeekdayOfTerm) {
        firstWeekdayOfCourse = weekdayNumber;
        break;
      }
    }

    let offset;
    if (firstWeekdayOfCourse >= startWeekdayOfTerm) {
      offset = firstWeekdayOfCourse - startWeekdayOfTerm;
    } else {
      offset = firstWeekdayOfCourse - startWeekdayOfTerm + 7;
    }

    this.startDate = addDays(termStartDate, offset);
    // Set every class on this end date because calendars will handle the offset
    this.endDate = termEndDate;
  }

  /**
   * Parses MyU calendar HTMl into UMNClasses with all relevant information.
   *
   * @param htmlSnippet HTML string containing the MyU calendar.
   * @param termStartDate start date of the term/semester.
   * @param termEndDate end date of the term/semester.
   * @returns array of UMN classes parsed from the calendar.
   */
  public static parseUMNClassesFromHTML(
    htmlSnippet: string,
    termStartDate: Date,
    termEndDate: Date
  ): Array<UMNClass> {
    const parser = new DOMParser();
    const htmlDoc = parser.parseFromString(htmlSnippet, 'text/html');
    const maybeCalendar = htmlDoc.getElementsByClassName('myu_calendar');
    if (maybeCalendar.length < 1) {
      throw new Error('Calendar could not be found');
    }
    const calendar = maybeCalendar[0];
    const days = calendar.getElementsByClassName('myu_calendar-day');
    const courseMap = new Map<string, UMNClass>();

    for (const day of days) {
      const dayName = day.getAttribute('data-day')?.toUpperCase();
      if (!dayName) continue;
      const courseElements = day.getElementsByClassName('myu_calendar-class');

      // Iterate through all classes in the day
      for (const courseElement of courseElements) {
        const c = new UMNClass();
        c.daysOfWeek.push(dayName);

        // Get the course info from the title
        const courseTitle =
          courseElement.getElementsByClassName('myu_calendar-class-name')[0].textContent;
        if (!courseTitle) continue;
        const courseNameAndSection = courseTitle.split(' ');
        c.dept = courseNameAndSection[0];
        c.courseNum = courseNameAndSection[1];
        c.section = courseNameAndSection[2];
        c.name = courseNameAndSection.slice(3).join(' ');

        const details = courseElement.getElementsByClassName('myu_calendar-class-details')[0]
          .textContent;
        if (!details) continue;
        const detailWords = details
          .split('\n')
          .flatMap((line) => line.split(' '))
          .filter((word) => word !== '' && word !== '-');

        // Append the type of time slot to the name
        c.name += ` ${detailWords[0]}`;

        // Extract the times
        const amOrPm = detailWords[3];
        let startTime = detailWords[1];
        let endTime = detailWords[2];
        if (amOrPm.toUpperCase() === 'PM' && getHours(startTime) > getHours(endTime)) {
          // Set the start time to AM if the end time is PM and startTime is greater than endTime
          startTime += 'AM';
          endTime += 'PM';
        } else {
          // Otherwise, set both times to the same AM/PM value
          startTime += amOrPm;
          endTime += amOrPm;
        }
        c.location = detailWords.slice(4).join(' ');
        c.startTimeStr = startTime;
        c.endTimeStr = endTime;

        if (!courseMap.has(c.hashCode())) {
          // Add the course to the map if it isn't already there
          courseMap.set(c.hashCode(), c);
        } else {
          // Add the day of the week to the class if it is already in the set
          courseMap.get(c.hashCode())?.daysOfWeek.push(dayName);
        }
      }
    }

    // Set the true course start and end dates
    for (const [, course] of courseMap.entries()) {
      course.setTrueCourseDates(termStartDate, termEndDate);
    }

    // Turn the map into an array and return
    return [...courseMap.values()];
  }
}
